### DISCUSSION TITLE
No idea how to setup livekit

### DISCUSSION THREAD

**Discussion by Sphinx2k** [bekriebel/fvtt-module-avclient-livekit#12] [Q&A] [answered]
I put in countless hours of figuring out how to get livekit running on my server. But as more and more problem are showing with the Jitsi foundry module, I tried again to get it running. After 2 hours and zero progress, I hope that I can get a hint on how to learn how to do it.
I'm a Linux amateur, with a good howto i can get most things working. So I have my foundry running, with an apache2 proxy and a jitsi-meet and the certificates are automatically renewing with lets-encrypt on an ubuntu 20LTS virtual private server.
But now I can't figure out how to setup livekit. The deployment docs are for cases that don't work for me. Deploy to VM would try to install caddy as webserver. I'm sure I would not work with apache2 already listening on most ports. 
Switching everything to caddy would mean I need to figure out how to setup foundry on caddy + let's encrypt. Without a good tutorial, this would also end in a situation that I could not resolve. 

Does anyone have any tips where I could start to learn how to get it working myself without a tutorial?

**Comment by Runesmite** [comment]
I'm right there with you. I'm using Docker and following the instructions but it seems like there's a big piece of information missing.

It's incredibly unclear whether or not Docker alone is suitable for running Livekit, or if it MUST be deployed into a VM. I'm able to get it working up to a point. Everything seems to work locally, but anyone connecting from the outside in are getting a "Timeout" error that I cannot find no matter how hard I search the web.

A more in-depth tutorial would be greatly appreciated.

In regards to you using Foundry and caddy+let's encrypt. I must say that is completely worth the switch. I am running my Foundry in a docker container and using caddy as a reverse proxy with portainer as a GUI and it is wonderful, I wish I had done that sooner. There is a good deal of documentation available for setting up Foundry in Docker, so hopefully that helps!

If you happen to make any breakthroughs on getting your LiveKit server to work, I'd love to hear them, and if I come across anything I'll try to remember to report back here as well! Good luck!

**Comment by bekriebel** [comment, marked as answer]
I agree that more guides will be useful, but for the module and for LiveKit itself. I hope to start making some in the near future but haven't had time available to do so.

I did enable a Wiki on this repo and a couple of great folks have contributed some guide on how to do deployments. This doesn't fit the exact scenarios you've mentioned, but perhaps it can help:

https://github.com/bekriebel/fvtt-module-avclient-livekit/wiki/Installing-LiveKit-on-an-Existing-Self-Hosted-Foundry-Server

The first thing I'd suggest checking for timeout issues is that the UDP ports are exposed and forwarded to the docker contain (and through your network). I also recommend testing your server setup with https://livekit.io/connection-test - this will help to tell you where in the connection process failures are happening.

I also suggest leaving TURN disabled to start with. It isn't needed for most users and it's better to get the system working without it before trying to add it.

I also plan on opening up my Discord to other users of the module. It's currently enabled for Patreon supporters only, but I'd like to have a place where folks can ask questions more easily. I need to get some moderation in place before I enable public access to the server, but I'll add info to the README when that is ready.

**Reply by Runesmite** [reply]
In my case, it was as simple as setting `use_external_ip: true`. I was running off the original `livekit.yaml`, and the original false that was used for testing was enabled. I wasn't "deploying" (at least in my mind) to a VM or Cloud service, but it makes sense to think of my container and server as a cloud service in hindsight.

For anyone else experiencing issues, definitely look at the config file examples given on livekit's own Deployment documentation.

Thank you for the Wiki link! I'll have to keep an eye on it.

**Reply by Sphinx2k** [reply]
I think I have it running, i will see next tuesday if it works for real. 
I startet with the wiki entry bekriebel mentioned.
https://github.com/bekriebel/fvtt-module-avclient-livekit/wiki/Installing-LiveKit-on-an-Existing-Self-Hosted-Foundry-Server
and this: https://foundryvtt.wiki/en/setup/hosting/Ubuntu-VM
That made the switch from apache2 to nginx as proxy easier than i thought. Even Let's Encrypt and the certificates work with little to no problem. 
At his point in the Wiki it stopped working "Install LiveKit", i could not connect to livekit not from the test nor foundry. I think the problem was that the livekit docker thing was trying to start the caddy server, witch failed because the ports are already used by nginx. So because i did not know how to fix it i went back to this test install
https://docs.livekit.io/guides/getting-started
setting the livekit.yaml to "use_external_ip: true" foundry did connect. 
So i changed the command i typed: "docker run --rm -p 7880:7880     -p 7881:7881     -p 7882:7882/udp     -v $PWD/livekit.yaml:/livekit.yaml     livekit/livekit-server     --config /livekit.yaml     --node-ip x.x.x.x" 
command to 
docker run -d --restart=always -p 7880:7880     -p 7881:7881     -p 7882:7882/udp     -v $PWD/livekit.yaml:/livekit.yaml     livekit/livekit-server     --config /livekit.yaml     --node-ip x.x.x.x" 
it even works after a restart of the whole Server.
-rm  -> remove docker after exit
-d -> as daemon or service i think
--restart=always -> restart when container crashed or not running

So i hope nothing breaks over time and maybe this is something that helps someone else.

**Comment by Jefferydo** [comment]
You've got a lot of nerve selling a technical product with such shitty documentation.

**Reply by bekriebel** [reply]
You mean the completly free and open source module maintained with my spare time? Yep... Lotta nerve.

The only part I "sell" is access to a managed multi-region server cluster for those who don't want to run their own servers. I charge a nominal fee to cover my costs and help sponser a bit of this modules development, but I assure you that it comes nowhere near covering the time and effort that has gone into this.

If you don't like how I've done this, there's an easy solution: don't use it. With comments like this, I'd rather you didn't anyway.
